### PR TITLE
move readable inside the loop of packets

### LINKED ISSUE #27917: `test_recv_mmsg_multi_iter_timeout` fails sometimes
#### Problem

I found out that from time to time `test_recv_mmsg_multi_iter_timeout` fails.
It happens because number of received packets is not equal to 32: https://github.com/solana-labs/solana/blob/master/streamer/src/nonblocking/recvmmsg.rs#L136

Depending on the semantics of function `recv_mmsg`, this failure might be due to:
1. mistake in test -- if by design we cannot guarantee that we received all the packets that we've sent when call `recv_mmsg`
2. mistake in the `recv_mmsg` implementation. As far as I understand there are 2 aspects: (i) `readiness` might be false positive, (ii) it should be checked for each of the packets in the loop. This might be resolve by using `recv_from` or by moving `readable` call inside the loop: see https://github.com/solana-labs/solana/pull/27916
3. Or it might be platform-dependent (I've reproduced only on my mac so far)

#### Proposed Solution

https://github.com/solana-labs/solana/pull/27916

### PR DESCRIPTION
**Pull request by KirillLykov**
#### Problem

We use `recv_mmsg` which are opportunistically try to read as many packets as possible from socket.
At the same time, we expect in tests that `recv_mmsg` reads exact number of packets. 

#### Summary of Changes

This PR proposes new function `recv_mmsg_exact` that guarantees to read N packets from socket.
Background: API of tokio usually provides one function without guarantee that it read all data (aka `read`) and another which gives such guarantee (aka `read_exact`) (thanks @alessandrod for pointing this out).
This PR is to resolve https://github.com/solana-labs/solana/issues/27917

### CODE REVIEW

File: streamer/src/nonblocking/recvmmsg.rs [hunk 1]
```diff
@@ -36,6 +38,21 @@ pub async fn recv_mmsg(
...
+/// Reads the exact number of packets required to fill `packets`
+pub async fn recv_mmsg_exact(
+    socket: &UdpSocket,
+    packets: &mut [Packet],
+) -> io::Result</*num packets:*/ usize> {
+    let total = packets.len();
+    let mut remaining = total;
+    while remaining != 0 {
+        let first = total - remaining;
+        let res = recv_mmsg(socket, &mut packets[first..]).await?;
```
Comment (ryleung-solana) [inline streamer/src/nonblocking/recvmmsg.rs:50]: This looks wrong if recv_mmsg doesn't receive all the packets on the first call...

Comment (ryleung-solana) [reply streamer/src/nonblocking/recvmmsg.rs:50]: Whoops, did not have enough caffeine, didn't see line 49.

File: streamer/src/nonblocking/recvmmsg.rs [hunk 2]
```diff
@@ -36,6 +38,21 @@ pub async fn recv_mmsg(
...
+        remaining -= res;
+    }
+    Ok(total)
```
Comment (ryleung-solana) [inline streamer/src/nonblocking/recvmmsg.rs:53]: Nit: this looks like it's always going to be Ok(packets.len()) if this function succeeds... Do we want this or should we either get rid of the usize, or return the number of packets read on error as well?

Comment (KirillLykov) [reply streamer/src/nonblocking/recvmmsg.rs:53]: I was thinking about mimicking behavior of https://docs.rs/tokio/latest/tokio/io/trait.AsyncReadExt.html#method.read_exact
I briefly looked into their implementation and it looks like they always return all or nothing: https://docs.rs/tokio/latest/src/tokio/io/util/read_exact.rs.html#65
I think `total => packets.len()` might be fine

**Review by ryleung-solana — APPROVED** [review]
LGTM but Jon should have a look first too.

**Review by joncinque — APPROVED** [review]
Looks great, thanks for looking into this!

**Pull request merged**